java/src/main/java/com/ncc/aif/ThreadSafeHasShapeFunction.java

```diff
@@ -32,10 +33,14 @@
  * @author Edward Curley
  */
 public class ThreadSafeHasShapeFunction extends HasShapeFunction {
-    private static ThreadLocal<Boolean> recursionIsErrorFlag = new ThreadLocal<Boolean>();
+    private static ThreadLocal<Boolean> recursionIsErrorFlag = new ThreadLocal<>();
 
     @Override
     protected NodeValue exec(Node focusNode, Node shapeNode, Node recursionIsError, FunctionEnv env) {
+        return exec(focusNode, shapeNode, recursionIsError, env.getActiveGraph(), DatasetImpl.wrap(env.getDataset()));
+    }
+
+    public static NodeValue exec(Node focusNode, Node shapeNode, Node recursionIsError, Graph activeGraph, Dataset dataset) {
```

Comment:
Good catch.  Would you object to me changing the protected method to invoke `ThreadSafeHasShapeFunction.exec()` instead?  If there is a downside to that, then I'm happy to take your suggestion as-is.

Reply:
As the static method isn't used anywhere but in the protected method, we should just not have the static method. Let me know if you have a reason for keeping it.

Reply:
Because `ThreadSafeHasShapeFunction` says it's including most of the code from
`HasShapeFunction`, and `HasShapeFunction` has is set up that way (with a protected wrapper around a static public method).  But I don't feel strongly about it.

Reply:
@ecurley-nextcentury advocates the more conservative approach of not including the static method and only including the minimum to get the private functionality changed.  Implementing this change.